# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLExtensions.h
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGLExtensions.h
--- a/dom/canvas/WebGLExtensions.h	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGLExtensions.h	Wed Jan 08 22:19:14 2020 +0000
@@ -34,13 +34,10 @@
 class WebGLTexture;
 class WebGLVertexArray;
 
-class WebGLExtensionBase : public nsWrapperCache,
-                           public WebGLContextBoundObject {
+class WebGLExtensionBase : public WebGLContextBoundObject<WebGLExtensionBase> {
  public:
   explicit WebGLExtensionBase(WebGLContext* webgl);
 
-  WebGLContext* GetParentObject() const { return mContext; }
-
   void MarkLost();
 
   void SetExplicit() {
@@ -50,31 +47,18 @@
 
   bool IsExplicit() { return mIsExplicit; }
 
-  NS_INLINE_DECL_CYCLE_COLLECTING_NATIVE_REFCOUNTING(WebGLExtensionBase)
-  NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_NATIVE_CLASS(WebGLExtensionBase)
+  NS_INLINE_DECL_REFCOUNTING(WebGLExtensionBase)
 
  protected:
   virtual ~WebGLExtensionBase();
 
   virtual void OnMarkLost() {}
-
-  bool mIsLost;
-
   virtual void OnSetExplicit() {}
 
-  bool mIsExplicit;
+  bool mIsLost = false;
+  bool mIsExplicit = false;
 };
 
-#define DECL_WEBGL_EXTENSION_GOOP             \
-  virtual JSObject* WrapObject(JSContext* cx, \
-                               JS::Handle<JSObject*> givenProto) override;
-
-#define IMPL_WEBGL_EXTENSION_GOOP(WebGLExtensionType, WebGLBindingType)        \
-  JSObject* WebGLExtensionType::WrapObject(JSContext* cx,                      \
-                                           JS::Handle<JSObject*> givenProto) { \
-    return dom::WebGLBindingType##_Binding::Wrap(cx, this, givenProto);        \
-  }
-
 ////
 
 class WebGLExtensionCompressedTextureASTC : public WebGLExtensionBase {
@@ -82,11 +66,9 @@
   explicit WebGLExtensionCompressedTextureASTC(WebGLContext* webgl);
   virtual ~WebGLExtensionCompressedTextureASTC();
 
-  void GetSupportedProfiles(dom::Nullable<nsTArray<nsString> >& retval) const;
+  Maybe<nsTArray<nsString>> GetSupportedProfiles() const;
 
   static bool IsSupported(const WebGLContext* webgl);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureBPTC final : public WebGLExtensionBase {
@@ -94,32 +76,24 @@
   explicit WebGLExtensionCompressedTextureBPTC(WebGLContext* webgl);
 
   static bool IsSupported(const WebGLContext* webgl);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureES3 : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionCompressedTextureES3(WebGLContext*);
   virtual ~WebGLExtensionCompressedTextureES3();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureETC1 : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionCompressedTextureETC1(WebGLContext*);
   virtual ~WebGLExtensionCompressedTextureETC1();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTexturePVRTC : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionCompressedTexturePVRTC(WebGLContext*);
   virtual ~WebGLExtensionCompressedTexturePVRTC();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureRGTC final : public WebGLExtensionBase {
@@ -127,8 +101,6 @@
   explicit WebGLExtensionCompressedTextureRGTC(WebGLContext* webgl);
 
   static bool IsSupported(const WebGLContext* webgl);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureS3TC : public WebGLExtensionBase {
@@ -137,8 +109,6 @@
   virtual ~WebGLExtensionCompressedTextureS3TC();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionCompressedTextureS3TC_SRGB : public WebGLExtensionBase {
@@ -147,16 +117,12 @@
   virtual ~WebGLExtensionCompressedTextureS3TC_SRGB();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionDebugRendererInfo : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionDebugRendererInfo(WebGLContext*);
   virtual ~WebGLExtensionDebugRendererInfo();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionDebugShaders : public WebGLExtensionBase {
@@ -164,10 +130,7 @@
   explicit WebGLExtensionDebugShaders(WebGLContext*);
   virtual ~WebGLExtensionDebugShaders();
 
-  void GetTranslatedShaderSource(const WebGLShader& shader,
-                                 nsAString& retval) const;
-
-  DECL_WEBGL_EXTENSION_GOOP
+  nsString GetTranslatedShaderSource(const WebGLShader& shader) const;
 };
 
 class WebGLExtensionDepthTexture : public WebGLExtensionBase {
@@ -176,16 +139,12 @@
   virtual ~WebGLExtensionDepthTexture();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionElementIndexUint : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionElementIndexUint(WebGLContext*);
   virtual ~WebGLExtensionElementIndexUint();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionExplicitPresent : public WebGLExtensionBase {
@@ -195,8 +154,6 @@
   static bool IsSupported(const WebGLContext*);
 
   void Present() const;
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionEXTColorBufferFloat : public WebGLExtensionBase {
@@ -205,8 +162,6 @@
   virtual ~WebGLExtensionEXTColorBufferFloat() {}
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionFBORenderMipmap : public WebGLExtensionBase {
@@ -215,8 +170,6 @@
   virtual ~WebGLExtensionFBORenderMipmap();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionFloatBlend : public WebGLExtensionBase {
@@ -225,8 +178,6 @@
   virtual ~WebGLExtensionFloatBlend();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionFragDepth : public WebGLExtensionBase {
@@ -235,8 +186,6 @@
   virtual ~WebGLExtensionFragDepth();
 
   static bool IsSupported(const WebGLContext* context);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionLoseContext : public WebGLExtensionBase {
@@ -246,8 +195,6 @@
 
   void LoseContext();
   void RestoreContext();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionSRGB : public WebGLExtensionBase {
@@ -256,16 +203,12 @@
   virtual ~WebGLExtensionSRGB();
 
   static bool IsSupported(const WebGLContext* context);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionStandardDerivatives : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionStandardDerivatives(WebGLContext*);
   virtual ~WebGLExtensionStandardDerivatives();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionShaderTextureLod : public WebGLExtensionBase {
@@ -274,16 +217,12 @@
   virtual ~WebGLExtensionShaderTextureLod();
 
   static bool IsSupported(const WebGLContext* context);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionTextureFilterAnisotropic : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionTextureFilterAnisotropic(WebGLContext*);
   virtual ~WebGLExtensionTextureFilterAnisotropic();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionTextureFloat : public WebGLExtensionBase {
@@ -294,16 +233,12 @@
   virtual ~WebGLExtensionTextureFloat();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionTextureFloatLinear : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionTextureFloatLinear(WebGLContext*);
   virtual ~WebGLExtensionTextureFloatLinear();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionTextureHalfFloat : public WebGLExtensionBase {
@@ -314,16 +249,12 @@
   virtual ~WebGLExtensionTextureHalfFloat();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionTextureHalfFloatLinear : public WebGLExtensionBase {
  public:
   explicit WebGLExtensionTextureHalfFloatLinear(WebGLContext*);
   virtual ~WebGLExtensionTextureHalfFloatLinear();
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionColorBufferFloat : public WebGLExtensionBase {
@@ -334,10 +265,7 @@
   static bool IsSupported(const WebGLContext*);
 
   void SetRenderable(const webgl::FormatRenderableState);
-
   void OnSetExplicit() override;
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionColorBufferHalfFloat : public WebGLExtensionBase {
@@ -348,10 +276,7 @@
   static bool IsSupported(const WebGLContext*);
 
   void SetRenderable(const webgl::FormatRenderableState);
-
   void OnSetExplicit() override;
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionDrawBuffers : public WebGLExtensionBase {
@@ -359,11 +284,9 @@
   explicit WebGLExtensionDrawBuffers(WebGLContext*);
   virtual ~WebGLExtensionDrawBuffers();
 
-  void DrawBuffersWEBGL(const dom::Sequence<GLenum>& buffers);
+  void DrawBuffersWEBGL(const nsTArray<GLenum>& buffers);
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionVertexArray : public WebGLExtensionBase {
@@ -375,8 +298,6 @@
   void DeleteVertexArrayOES(WebGLVertexArray* array);
   bool IsVertexArrayOES(const WebGLVertexArray* array);
   void BindVertexArrayOES(WebGLVertexArray* array);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionInstancedArrays : public WebGLExtensionBase {
@@ -391,8 +312,6 @@
   void VertexAttribDivisorANGLE(GLuint index, GLuint divisor);
 
   static bool IsSupported(const WebGLContext* webgl);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionBlendMinMax : public WebGLExtensionBase {
@@ -401,8 +320,6 @@
   virtual ~WebGLExtensionBlendMinMax();
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionDisjointTimerQuery : public WebGLExtensionBase {
@@ -416,14 +333,11 @@
   void BeginQueryEXT(GLenum target, WebGLQuery& query) const;
   void EndQueryEXT(GLenum target) const;
   void QueryCounterEXT(WebGLQuery& query, GLenum target) const;
-  void GetQueryEXT(JSContext* cx, GLenum target, GLenum pname,
-                   JS::MutableHandleValue retval) const;
-  void GetQueryObjectEXT(JSContext* cx, const WebGLQuery& query, GLenum pname,
-                         JS::MutableHandleValue retval) const;
+  MaybeWebGLVariant GetQueryEXT(GLenum target, GLenum pname) const;
+  MaybeWebGLVariant GetQueryObjectEXT(const WebGLQuery& query,
+                                      GLenum pname) const;
 
   static bool IsSupported(const WebGLContext*);
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 class WebGLExtensionMOZDebug final : public WebGLExtensionBase {
@@ -431,11 +345,77 @@
   explicit WebGLExtensionMOZDebug(WebGLContext* webgl);
   virtual ~WebGLExtensionMOZDebug();
 
-  void GetParameter(JSContext* cx, GLenum pname,
-                    JS::MutableHandle<JS::Value> retval, ErrorResult& er) const;
+  MaybeWebGLVariant GetParameter(GLenum pname) const;
+};
+
+template <WebGLExtensionID ext>
+struct WebGLExtensionClassMap;
+
+#define DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(_extensionID, _extensionClass) \
+  template <>                                                                 \
+  struct WebGLExtensionClassMap<WebGLExtensionID::_extensionID> {             \
+    using Type = _extensionClass;                                             \
+  };
 
-  DECL_WEBGL_EXTENSION_GOOP
-};
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(ANGLE_instanced_arrays,
+                                       WebGLExtensionInstancedArrays)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_blend_minmax,
+                                       WebGLExtensionBlendMinMax)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_color_buffer_float,
+                                       WebGLExtensionColorBufferFloat)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_color_buffer_half_float,
+                                       WebGLExtensionColorBufferHalfFloat)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_texture_compression_bptc,
+                                       WebGLExtensionCompressedTextureBPTC)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_texture_compression_rgtc,
+                                       WebGLExtensionCompressedTextureRGTC)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_frag_depth, WebGLExtensionFragDepth)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_sRGB, WebGLExtensionSRGB)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_shader_texture_lod,
+                                       WebGLExtensionShaderTextureLod)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_texture_filter_anisotropic,
+                                       WebGLExtensionTextureFilterAnisotropic)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(EXT_disjoint_timer_query,
+                                       WebGLExtensionDisjointTimerQuery)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(MOZ_debug, WebGLExtensionMOZDebug)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_element_index_uint,
+                                       WebGLExtensionElementIndexUint)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_standard_derivatives,
+                                       WebGLExtensionStandardDerivatives)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_texture_float,
+                                       WebGLExtensionTextureFloat)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_texture_float_linear,
+                                       WebGLExtensionTextureFloatLinear)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_texture_half_float,
+                                       WebGLExtensionTextureHalfFloat)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_texture_half_float_linear,
+                                       WebGLExtensionTextureHalfFloatLinear)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(OES_vertex_array_object,
+                                       WebGLExtensionVertexArray)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_color_buffer_float,
+                                       WebGLExtensionEXTColorBufferFloat)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_astc,
+                                       WebGLExtensionCompressedTextureASTC)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_etc,
+                                       WebGLExtensionCompressedTextureES3)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_etc1,
+                                       WebGLExtensionCompressedTextureETC1)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_pvrtc,
+                                       WebGLExtensionCompressedTexturePVRTC)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_s3tc,
+                                       WebGLExtensionCompressedTextureS3TC)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_compressed_texture_s3tc_srgb,
+                                       WebGLExtensionCompressedTextureS3TC_SRGB)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_debug_renderer_info,
+                                       WebGLExtensionDebugRendererInfo)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_debug_shaders,
+                                       WebGLExtensionDebugShaders)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_depth_texture,
+                                       WebGLExtensionDepthTexture)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_draw_buffers,
+                                       WebGLExtensionDrawBuffers)
+DEFINE_WEBGL_EXTENSION_CLASS_MAP_ENTRY(WEBGL_lose_context,
+                                       WebGLExtensionLoseContext)
 
 class WebGLExtensionMultiview : public WebGLExtensionBase {
  public:
@@ -447,8 +427,6 @@
                                       WebGLTexture* texture, GLint level,
                                       GLint baseViewIndex,
                                       GLsizei numViews) const;
-
-  DECL_WEBGL_EXTENSION_GOOP
 };
 
 }  // namespace mozilla